_posts/2016-12-27-angular-2-advance-testing-with-custom-matchers.md

```diff
@@ -0,0 +1,650 @@
+// 'MyComponent' construction requires MatchMedia parameter
+
+import { MyComponent } from './viewer/MyComponent';
+
+// Configure testbed to register components and prepare services
+beforeEach(() => {
+
+  TestBed.configureTestingModule({
+    imports: [ FlexLayoutModule.forRoot() ],
+    declarations: [MyComponent],
+    providers: [
+      BreakPointRegistry,
+      BreakPointsProvider,
+      {
+        provide: MatchMedia,
+        useClass: MockMatchMedia
+      }
+    ]
+  })
+
+});
+// Helper function to easily build a component Fixture using the specified template
+function createTestComponent(template: string): ComponentFixture<Type<any>>
+{
+  return TestBed
+    .overrideComponent(MyComponent, {set: {template: template}} )
+    .createComponent(MyComponent);
+}
+{% endraw %}
+{% endhighlight %}
+
+Before each test, I want configure a new, fresh testing module with <u>only</u> the providers, 
+components, and directives I need for this test module. And I use `createTestComponent()` to easily
+specify a custom template that will be used when instantiating a component.
+
+At first, this complexity may seem like overkill. But let's consider two critical requirements 
+shown in the sample above:
+
+* `MatchMedia` instantiation requires an injected `BreakPointRegistry` instance
+* `MyComponent` instantiation requires an injected `MatchMedia` instance
+
+Even with these requirements, developers should NOT have to worry about all these internals just 
+to test **MyComponent**. Using ngModule, DI, and Angular... I now don't have to worry about those 
+details.
+
+This is just like those real-world scenarios where your components, directives, and services will 
+have complex dependencies upon providers and non-trivial construction processes. 
+
+And this is where **TestBed** demonstrates its real value!
+
+> I am not using external templates nor any other resources or services that are asynchronous. 
+So I do not discuss the `async()` wrapper function: since I do not need to asynchronously prepare 
+the TestBed nor `compileComponents()`.
+
+### 1) Testing Angular Directives
+
+With relative ease, developers can find literature on testing Angular Services and Components. 
+Yet the <u>How-to's for testing Directives</u> is oddly not well documented.
+
+
+Unlike Components with their associated templates, Directives do not have templates. This means 
+that I cannot simply **import** a Directive and create it. 
+
+Yet the solution is rather easy! I need to:
+
+* configure a TestBed that imports and declares the directive(s),
+* prepare a shell test Component, and
+* use a custom template which uses the Directive selector(s) as attribute(s).
+
+
+Since Directives usually affect the host element, our tests will check if those changes to the host 
+element are present and valid.
+
+###### Testing the fxLayout Directive
+
+I have found that real-world exmaples provide the best examples for solutions and reusable techniques.
```

Comment:
Instead of "I have found that....", rather sth. like "We believe that..."

and maybe instead of "the best", something like: "provide better ... than ...."

Maybe it's just me but I try to avoid saying that something is "the best/strongest/hardest/easiest" because it often depends on the person how she deals with things.